File: src/main/java/com/mercadez/service/ProdutoService.java
package com.mercadez.service;

import com.mercadez.dto.*;
import com.mercadez.exception.*;
import com.mercadez.model.Produto;
import com.mercadez.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ProdutoService {

    private final ProdutoRepository  produtoRepo;
    private final AfiliadoRepository afiliadoRepo;

    public ProdutoService(ProdutoRepository produtoRepo, AfiliadoRepository afiliadoRepo) {
        this.produtoRepo  = produtoRepo;
        this.afiliadoRepo = afiliadoRepo;
    }

    /** Cadastro de produto — afiliadoId vem do JWT */
    @Transactional
    public ProdutoResponse cadastrar(CadastroProdutoRequest req, Long afiliadoId) {
        var afiliado = afiliadoRepo.findById(afiliadoId)
                .orElseThrow(() -> new NaoEncontradoException("Afiliado não encontrado."));

        Produto p = Produto.builder()
                .nomeProduto(req.getNomeProduto())
                .tags(req.getTags())
                .preco(req.getPreco())
                .quantidade(req.getQuantidade())
                .afiliado(afiliado)
                .build();

        return ProdutoResponse.de(produtoRepo.save(p));
    }

    /** Lista todos os produtos ativos (página pública) */
    @Transactional(readOnly = true)
    public List<ProdutoResponse> listarTodos() {
        return produtoRepo.findByAtivoTrue()
                .stream().map(ProdutoResponse::de).toList();
    }

    /** Lista produtos de um afiliado específico */
    @Transactional(readOnly = true)
    public List<ProdutoResponse> listarPorAfiliado(Long afiliadoId) {
        return produtoRepo.findByAfiliadoIdAndAtivoTrue(afiliadoId)
                .stream().map(ProdutoResponse::de).toList();
    }

    /** Busca por nome ou tag (barra de pesquisa do frontend) */
    @Transactional(readOnly = true)
    public List<ProdutoResponse> buscar(String termo) {
        if (termo == null || termo.isBlank()) return listarTodos();
        return produtoRepo.buscarPorTermo(termo)
                .stream().map(ProdutoResponse::de).toList();
    }

    /** Comparação de preços do mesmo produto em vários mercados */
    @Transactional(readOnly = true)
    public List<ProdutoResponse> comparar(String nome) {
        if (nome == null || nome.isBlank())
            throw new NegocioException("Informe o nome do produto para comparar.");
        return produtoRepo.compararPrecos(nome)
                .stream().map(ProdutoResponse::de).toList();
    }

    /** Atualizar produto — verifica dono */
    @Transactional
    public ProdutoResponse atualizar(Long id, CadastroProdutoRequest req, Long afiliadoId) {
        Produto p = produtoRepo.findById(id)
                .orElseThrow(() -> new NaoEncontradoException("Produto não encontrado."));

        if (!p.getAfiliado().getId().equals(afiliadoId)) {
            throw new NegocioException("Você não tem permissão para editar este produto.");
        }

        p.setNomeProduto(req.getNomeProduto());
        p.setTags(req.getTags());
        p.setPreco(req.getPreco());
        p.setQuantidade(req.getQuantidade());

        return ProdutoResponse.de(produtoRepo.save(p));
    }

    /** Soft-delete — verifica dono */
    @Transactional
    public void deletar(Long id, Long afiliadoId) {
        Produto p = produtoRepo.findById(id)
                .orElseThrow(() -> new NaoEncontradoException("Produto não encontrado."));

        if (!p.getAfiliado().getId().equals(afiliadoId)) {
            throw new NegocioException("Você não tem permissão para excluir este produto.");
        }

        p.setAtivo(false);
        produtoRepo.save(p);
    }
}

File: app/services/produto_service.py
"""Equivalente Python de service/ProdutoService.java"""
from sqlalchemy import select, or_
from sqlalchemy.orm import Session, joinedload

from app.exceptions import AcessoNegadoException, NaoEncontradoException, NegocioException
from app.models.afiliado import Afiliado
from app.models.produto import Produto
from app.schemas.produto import CadastroProdutoRequest, ProdutoResponse


def cadastrar(db: Session, req: CadastroProdutoRequest, afiliado_id: int) -> ProdutoResponse:
    afiliado = db.get(Afiliado, afiliado_id)
    if afiliado is None:
        raise NaoEncontradoException("Afiliado nao encontrado.")

    produto = Produto(
        nome_produto=req.nomeProduto,
        tags=req.tags,
        preco=req.preco,
        quantidade=req.quantidade,
        afiliado_id=afiliado_id,
    )
    db.add(produto)
    db.commit()
    db.refresh(produto)
    return ProdutoResponse.de(produto)


def listar_todos(db: Session) -> list[ProdutoResponse]:
    produtos = db.scalars(
        select(Produto).options(joinedload(Produto.afiliado)).where(Produto.ativo.is_(True))
    ).all()
    return [ProdutoResponse.de(p) for p in produtos]


def listar_por_afiliado(db: Session, afiliado_id: int) -> list[ProdutoResponse]:
    produtos = db.scalars(
        select(Produto)
        .options(joinedload(Produto.afiliado))
        .where(Produto.afiliado_id == afiliado_id, Produto.ativo.is_(True))
    ).all()
    return [ProdutoResponse.de(p) for p in produtos]


def buscar(db: Session, termo: str) -> list[ProdutoResponse]:
    if not termo or not termo.strip():
        return listar_todos(db)

    padrao = f"%{termo.lower()}%"
    produtos = db.scalars(
        select(Produto)
        .options(joinedload(Produto.afiliado))
        .where(
            Produto.ativo.is_(True),
            or_(
                Produto.nome_produto.ilike(padrao),
                Produto.tags.ilike(padrao),
            ),
        )
        .order_by(Produto.preco.asc())
    ).all()
    return [ProdutoResponse.de(p) for p in produtos]


def comparar(db: Session, nome: str) -> list[ProdutoResponse]:
    if not nome or not nome.strip():
        raise NegocioException("Informe o nome do produto para comparar.")

    padrao = f"%{nome.lower()}%"
    produtos = db.scalars(
        select(Produto)
        .options(joinedload(Produto.afiliado))
        .where(Produto.ativo.is_(True), Produto.nome_produto.ilike(padrao))
        .order_by(Produto.preco.asc())
    ).all()
    return [ProdutoResponse.de(p) for p in produtos]


def atualizar(db: Session, produto_id: int, req: CadastroProdutoRequest, afiliado_id: int) -> ProdutoResponse:
    produto = db.get(Produto, produto_id)
    if produto is None:
        raise NaoEncontradoException("Produto nao encontrado.")

    if produto.afiliado_id != afiliado_id:
        raise AcessoNegadoException("Voce nao tem permissao para editar este produto.")

    produto.nome_produto = req.nomeProduto
    produto.tags = req.tags
    produto.preco = req.preco
    produto.quantidade = req.quantidade

    db.commit()
    db.refresh(produto)
    return ProdutoResponse.de(produto)


def deletar(db: Session, produto_id: int, afiliado_id: int) -> None:
    produto = db.get(Produto, produto_id)
    if produto is None:
        raise NaoEncontradoException("Produto nao encontrado.")

    if produto.afiliado_id != afiliado_id:
        raise AcessoNegadoException("Voce nao tem permissao para excluir este produto.")

    produto.ativo = False  # soft-delete, igual ao backend Java
    db.commit()
